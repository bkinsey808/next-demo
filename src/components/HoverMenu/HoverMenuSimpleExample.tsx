import { Transition } from "@headlessui/react";
import clsx from "clsx";
import cntl from "cntl";
import { FC, Fragment, useCallback, useState } from "react";

import { HoverMenu } from "./HoverMenu";
import { MenuItemConfig } from "./types";

export interface ConfigData {
  label: string;
}

export const enum ColorEnum {
  GREEN = "green",
  YELLOW = "yellow",
  RED = "red",
}

export const ITEMS: ColorEnum[] = [
  ColorEnum.GREEN,
  ColorEnum.YELLOW,
  ColorEnum.RED,
];

export const ITEM_CONFIG: MenuItemConfig<ConfigData, ColorEnum> = {
  [ColorEnum.GREEN]: {
    label: "Green",
  },
  [ColorEnum.RED]: {
    label: "Red",
  },
  [ColorEnum.YELLOW]: {
    label: "Yellow",
  },
} as const;

export const HoverMenuSimpleExample: FC = () => {
  const [activeColor, setActiveColor] = useState<ColorEnum>(ColorEnum.GREEN);

  return (
    <HoverMenu<ConfigData, ColorEnum>
      activeKey={activeColor}
      items={ITEMS}
      itemConfig={ITEM_CONFIG}
      onSelect={setActiveColor}
      ariaLabel={"Select Color"}
      WrapperComponent={useCallback(
        ({ children }) => (
          <div className="relative inline-block">{children}</div>
        ),
        []
      )}
      MenuButtonComponent={useCallback(
        ({ activeKey }) => (
          <div
            className="
              inline-flex 
              justify-center
              rounded-md 
              bg-black 
              bg-opacity-20 
              px-4 py-2
              text-sm 
              font-medium 
              text-white 
              hover:bg-opacity-30                
              focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75
           "
          >
            {ITEM_CONFIG[activeKey].label}
          </div>
        ),
        []
      )}
      TransitionComponent={useCallback(
        ({ children, isActive }) => (
          <Transition
            show={isActive}
            as={Fragment}
            enter={cntl`transition ease-out duration-100`}
            enterFrom={cntl`transform opacity-0 scale-95`}
            enterTo={cntl`transform opacity-100 scale-100`}
            leave={cntl`transition ease-in duration-75`}
            leaveFrom={cntl`transform opacity-100 scale-100`}
            leaveTo={cntl`transform opacity-0 scale-95`}
          >
            {children}
          </Transition>
        ),
        []
      )}
      MenuItemsComponent={useCallback(
        ({ children }) => (
          <div
            className="
              absolute left-0
              z-10 
              flex
              w-64
              origin-top-right
              flex-col
              divide-y
              divide-gray-100
              rounded-md
              bg-white
              px-1
              py-1
              shadow-lg
              ring-1 ring-black ring-opacity-5
              focus:outline-none
            "
          >
            {children}
          </div>
        ),
        []
      )}
      MenuItemComponent={useCallback(
        ({ itemKey, itemConfig, active, selected }) => (
          <div
            className={clsx(
              `
              group
              flex
              w-full
              items-center
              rounded-md
              px-2
              py-2 text-sm
            `,
              {
                "bg-violet-500 text-white": active,
                "text-gray-900": !active,
              }
            )}
          >
            {itemConfig[itemKey].label}: {JSON.stringify({ active, selected })}
          </div>
        ),
        []
      )}
    />
  );
};
