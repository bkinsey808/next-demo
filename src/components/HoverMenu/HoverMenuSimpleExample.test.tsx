import { fireEvent, render, within } from "@testing-library/react";

import { HoverMenuSimpleExample } from "./HoverMenuSimpleExample";

describe("HoverMenuSimpleExample", () => {
  it("renders color", async () => {
    const { getByLabelText, getByText, asFragment } = render(
      <HoverMenuSimpleExample />
    );
    expect(asFragment()).toMatchSnapshot();

    const menuButton = await getByLabelText(/select color/i);
    expect(menuButton).toBeInTheDocument();

    // click doesn't work because of hover menu implementation
    fireEvent.mouseDown(menuButton);

    const redButton = await getByText(/Red/i);
    fireEvent.click(redButton);
    const { getByText: menuButtonGetByText } = within(menuButton);

    // Confirm color was selected. Will throw error if not found.
    menuButtonGetByText(/Red/i);
    expect(asFragment()).toMatchSnapshot();
  });
});
