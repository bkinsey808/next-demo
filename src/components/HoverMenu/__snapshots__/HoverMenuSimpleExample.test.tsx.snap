// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverMenuSimpleExample renders color 1`] = `
<DocumentFragment>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Select Color"
    class=" justify-left relative  inline-flex w-full rounded-md  bg-black  bg-opacity-20  px-4 py-2 text-sm  font-medium  text-white  hover:bg-opacity-30                 focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75 "
    id="headlessui-menu-button-1"
    type="button"
  >
    <div>
      Green
    </div>
  </button>
</DocumentFragment>
`;

exports[`HoverMenuSimpleExample renders color 2`] = `
<DocumentFragment>
  <button
    aria-controls="headlessui-menu-items-3"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Select Color"
    class=" justify-left relative  inline-flex w-full rounded-md  bg-black  bg-opacity-20  px-4 py-2 text-sm  font-medium  text-white  hover:bg-opacity-30                 focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75 "
    id="headlessui-menu-button-1"
    type="button"
  >
    <div>
      Red
    </div>
  </button>
  <div
    class="absolute transition ease-in duration-75 transform opacity-100 scale-100"
  >
    <div
      aria-activedescendant="headlessui-menu-item-4"
      aria-labelledby="headlessui-menu-button-1"
      class=" absolute left-0  z-10  mt-2  w-64  origin-top-right  divide-y  divide-gray-100  rounded-md  bg-white  shadow-lg  ring-1 ring-black ring-opacity-5  focus:outline-none "
      id="headlessui-menu-items-3"
      role="menu"
      tabindex="0"
    >
      <div
        class="px-1 py-1"
      >
        <button
          class="
              group 
              flex 
              w-full 
              items-center 
              rounded-md 
              px-2 
              py-2 text-sm 
              focus:outline-none focus-visible:outline-none
             bg-violet-500 text-white"
          id="headlessui-menu-item-4"
          role="menuitem"
          tabindex="-1"
        >
          <div>
            Green: false
          </div>
        </button>
        <button
          class="
              group 
              flex 
              w-full 
              items-center 
              rounded-md 
              px-2 
              py-2 text-sm 
              focus:outline-none focus-visible:outline-none
             text-gray-900"
          id="headlessui-menu-item-5"
          role="menuitem"
          tabindex="-1"
        >
          <div>
            Yellow: false
          </div>
        </button>
        <button
          class="
              group 
              flex 
              w-full 
              items-center 
              rounded-md 
              px-2 
              py-2 text-sm 
              focus:outline-none focus-visible:outline-none
             text-gray-900"
          id="headlessui-menu-item-6"
          role="menuitem"
          tabindex="-1"
        >
          <div>
            Red: false
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
